pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -728,6 +729,362 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
+	if err := scaleUpStatefulSetControl(set, ssc, spc, assertBurstInvariants); err != nil {
+		t.Fatal(err)
+	}
+	set, err := spc.setsLister.StatefulSets(set.Namespace).Get(set.Name)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return set, spc, ssc, maxUnavailable
+}
+
+func TestStatefulSetControlRollingUpdateWithMaxUnavailableInOrderedModeVerifyInvariant(t *testing.T) {
+	// Make all pods in statefulset unavailable one by one
+	// and verify that RollingUpdate doesnt proceed with maxUnavailable set
+	// this could have been a simple loop, keeping it like this to be able
+	// to add more params here.
+	testCases := []struct {
+		ordinalOfPodToTerminate []int
+	}{
+		{[]int{}},
+		{[]int{5}},
+		{[]int{3}},
+		{[]int{4}},
+		{[]int{5, 4}},
+		{[]int{5, 3}},
+		{[]int{4, 3}},
+		{[]int{5, 4, 3}},
+		//	{2}, // 2 will not be terminated because its less than partition
+	}
+	for _, tc := range testCases {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.MaxUnavailableStatefulSet, true)()
+		set, spc, ssc, maxUnavailable := setupForInvariant(t)
+		t.Run(fmt.Sprintf("terminating pod at ordinal %d", tc.ordinalOfPodToTerminate), func(t *testing.T) {
+			status := apps.StatefulSetStatus{Replicas: 6}
+			updateRevision := &apps.ControllerRevision{}
+
+			for i := 0; i < len(tc.ordinalOfPodToTerminate); i++ {
+				// Ensure at least one pod is unavailable before trying to update
+				_, err := spc.addTerminatingPod(set, tc.ordinalOfPodToTerminate[i])
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			originalPods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			sort.Sort(ascendingOrdinal(originalPods))
+
+			// start to update
+			set.Spec.Template.Spec.Containers[0].Image = "foo"
+
+			// try to update the statefulset
+			if _, err = updateStatefulSetAfterInvariantEstablished(context.TODO(), ssc.(*defaultStatefulSetControl), set, originalPods, updateRevision, status); err != nil {
+				t.Fatal(err)
+			}
+			pods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			sort.Sort(ascendingOrdinal(pods))
+
+			fmt.Println("pods already terminating ", len(tc.ordinalOfPodToTerminate))
+			fmt.Println("pods to update ", maxUnavailable.IntValue())
+			expecteddPodsToBeDeleted := maxUnavailable.IntValue() - len(tc.ordinalOfPodToTerminate)
+			if expecteddPodsToBeDeleted < 0 {
+				expecteddPodsToBeDeleted = 0
+			}
+			fmt.Println("pods to be deleted", expecteddPodsToBeDeleted)
+
+			expectedPodsAfterUpdate := 6 - expecteddPodsToBeDeleted
+
+			if len(pods) != expectedPodsAfterUpdate {
+				t.Errorf("Expected pods %v, got pods %v", expectedPodsAfterUpdate, len(pods))
+			}
+
+		})
+	}
+}
+
+func TestStatefulSetControlRollingUpdateWithMaxUnavailableInBurstMode(t *testing.T) {
```

Comment:
There is a lot of duplication between this method and `TestStatefulSetControlRollingUpdateWithMaxUnavailableInOrderedMode` - Can you make mode as an argument and make the test table driven?